codec/dns/codec.mbt

```diff
@@ -0,0 +1,279 @@
+  let opcode = OpCode::from_int(flags.lsr(11).land(0xF))
+  let aa = flags.land(0x0400) != 0
+  let tc = flags.land(0x0200) != 0
+  let rd = flags.land(0x0100) != 0
+  let ra = flags.land(0x0080) != 0
+  let z = flags.lsr(4).land(0x7)
+  let rcode = ResponseCode::from_int(flags.land(0xF))
+  
+  // Counts
+  let qdcount = data[4].to_int().lsl(8).lor(data[5].to_int())
+  let ancount = data[6].to_int().lsl(8).lor(data[7].to_int())
+  let nscount = data[8].to_int().lsl(8).lor(data[9].to_int())
+  let arcount = data[10].to_int().lsl(8).lor(data[11].to_int())
+  
+  Ok(Header::new(id, qr, opcode, aa, tc, rd, ra, z, rcode, qdcount, ancount, nscount, arcount))
+}
+
+///| Encode a DNS question to buffer
+pub fn encode_question(question : Question, buf : &@buffer.Buffer) -> Result[Unit, DnsError] {
+  // Encode domain name
+  match encode_domain_name(question.qname, buf) {
+    Ok(_) => ()
+    Err(e) => return Err(e)
+  }
+  
+  // Query type (16 bits)
+  let qtype = question.qtype.to_int()
+  buf.write_byte(qtype.lsr(8).land(0xFF).to_byte()) |> ignore
+  buf.write_byte(qtype.land(0xFF).to_byte()) |> ignore
+  
+  // Query class (16 bits)
+  let qclass = question.qclass.to_int()
+  buf.write_byte(qclass.lsr(8).land(0xFF).to_byte()) |> ignore
+  buf.write_byte(qclass.land(0xFF).to_byte()) |> ignore
+  
+  Ok(())
+}
+
+///| Decode a DNS question from buffer
+///| Returns (question, bytes_consumed)
+pub fn decode_question(data : FixedArray[Byte], offset : Int) -> Result[(Question, Int), DnsError] {
+  let name_result = decode_domain_name(data, offset)
+  match name_result {
+    Err(e) => return Err(e)
+    Ok((qname, name_len)) => {
+      let mut pos = offset + name_len
+      
+      if pos + 4 > data.length() {
+        return Err(BufferTooSmall("Not enough data for question"))
+      }
+      
+      // Query type
+      let qtype = RecordType::from_int(
+        data[pos].to_int().lsl(8).lor(data[pos + 1].to_int())
+      )
+      pos = pos + 2
+      
+      // Query class
+      let qclass = RecordClass::from_int(
+        data[pos].to_int().lsl(8).lor(data[pos + 1].to_int())
+      )
+      pos = pos + 2
+      
+      Ok((Question::new(qname, qtype, qclass), pos - offset))
+    }
+  }
+}
+
+///| Encode a complete DNS message to buffer
+pub fn encode_message(msg : Message) -> Result[&@buffer.Buffer, DnsError] {
+  let buf = @buffer.HeapByteBuf::new(512) // Standard DNS packet size
+  
+  // Encode header
+  encode_header(msg.header, buf as &@buffer.Buffer)
+  
+  // Encode questions
+  for question in msg.questions {
+    match encode_question(question, buf as &@buffer.Buffer) {
+      Ok(_) => ()
+      Err(e) => return Err(e)
+    }
+  }
+  
+  // For now, we only support encoding questions
+  // Answers, authority, and additional sections would be encoded similarly
+  
+  Ok(buf as &@buffer.Buffer)
+}
+
+///| Decode a complete DNS message from buffer
+pub fn decode_message(data : FixedArray[Byte]) -> Result[Message, DnsError] {
```

Comment:
Can we pass Buffer instead of FixedArray of Byte? Is there some API you cannot interact with Buffer.